### PR TITLE
OHRM5X-1416: Develop Candidate Schedule Interview API

### PR DESCRIPTION
**Pull request by SandeepaOHRM**
This PR contains:-
  - OHRM5X-1416: Candidate Schedule Interview API
  - OHRM5X-1423: Interview Attachment API

### CODE REVIEW

File: installer/Migration/V5_1_0/permission/api.yaml [hunk 1]
```diff
@@ -265,3 +265,29 @@ apiv2_recruitment_candidate_shortlist:
...
+  api: OrangeHRM\Recruitment\Api\InterviewAttachmentAPI
+  module: recruitment
+  allowed:
+    read: true
+    create: true
+    update: true
+    delete: false
+  permissions:
+    - { role: Admin, permission: { read: true, create: true, update: true, delete: false, self: false } }
+    - { role: HiringManager, permission: { read: true, create: true, update: true, delete: false, self: false } }
```
Comment (RajithaKumara) [inline installer/Migration/V5_1_0/permission/api.yaml:293]: can interviewer attach anything?

File: src/plugins/orangehrmRecruitmentPlugin/Api/CandidateScheduleInterviewAPI.php [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+            ) {
+                throw $this->getForbiddenException();
+            }
+
+            $numberOfInterviewsScheduled = $this->getCandidateService()
+                ->getCandidateDao()
+                ->getInterviewCountByCandidateId($candidateId);
+
+            if ($numberOfInterviewsScheduled >= self::MAXIMUM_ALLOWED_INTERVIEWS_COUNT) {
+                throw $this->getBadRequestException('You Can not Schedule More Than Two Interviews Per Candidate');
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/CandidateScheduleInterviewAPI.php:119]: double check message, if we are going to show in FE

File: src/plugins/orangehrmRecruitmentPlugin/Api/CandidateScheduleInterviewAPI.php [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    use AuthUserTrait;
+    use DateTimeHelperTrait;
+    use EntityManagerHelperTrait;
+
+    public const PARAMETER_CANDIDATE_ID = 'candidateId';
+    public const PARAMETER_INTERVIEW_NAME = 'interviewName';
+    public const PARAMETER_INTERVIEW_DATE = 'interviewDate';
+    public const PARAMETER_INTERVIEW_TIME = 'interviewTime';
+    public const PARAMETER_NOTE = 'note';
+    public const PARAMETER_INTERVIEWERS = 'interviewers';
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/CandidateScheduleInterviewAPI.php:62]:
```suggestion
    public const PARAMETER_INTERVIEWERS = 'interviewerEmpNumbers';
```
shall we go like this? followed in performance module also

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function getAll(): EndpointResult
+    {
+        $interviewId = $this->getRequestParams()->getInt(
+            RequestParams::PARAM_TYPE_ATTRIBUTE,
+            self::PARAMETER_INTERVIEW_ID
+        );
+        $attachments = $this->getRecruitmentAttachmentService()
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:64]: better to handle usual getAll with search filter param object :relaxed:

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+                RequestParams::PARAM_TYPE_BODY,
+                self::PARAMETER_COMMENT
+            )
+        );
+    }
+
+    /**
+     * @param InterviewAttachment $InterviewAttachment
+     */
+    private function setBase64Attachment(InterviewAttachment $InterviewAttachment): void
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:123]:
```suggestion
     * @param InterviewAttachment $interviewAttachment
     */
    private function setBase64Attachment(InterviewAttachment $interviewAttachment): void
```
refactor rename

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    {
+        throw $this->getNotImplementedException();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getOne(): EndpointResult
+    {
+        $attachmentId = $this->getRequestParams()->getInt(
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:191]: seems can anyone access any attachment by id

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+            ),
+        );
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function update(): EndpointResult
+    {
+        $attachmentId = $this->getRequestParams()->getInt(
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:221]: seems getOne & update both need permission check

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php (hunk 6 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:191]: yes, Once user roles implemented, we can use ACCECCIBLE_ENTITY_IDS validation I guess

Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:191]: yes, Once user roles implemented, we can use ACCECCIBLE_ENTITY_IDS validation I guess

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php (hunk 7 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:221]: same as above

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php (hunk 5 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:123]: done

File: src/plugins/orangehrmRecruitmentPlugin/Api/CandidateScheduleInterviewAPI.php (hunk 3 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/CandidateScheduleInterviewAPI.php:62]: done

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php (hunk 4 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:70]: done

File: src/plugins/orangehrmRecruitmentPlugin/Dto/InterviewAttachmentSearchFilterParams.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+{
+    /**
+     * @var int
+     */
+    protected int $interviewId;
+
+    public function __construct()
+    {
+        $this->setSortField('attachment.interviewId');
+        $this->setSortOrder(ListSorter::ASCENDING);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Dto/InterviewAttachmentSearchFilterParams.php:35]:
```suggestion
```
I think ascending is the default behaviour

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+                self::PARAMETER_INTERVIEW_ID
+            )
+        );
+        $attachments = $this->getRecruitmentAttachmentService()
+            ->getRecruitmentAttachmentDao()
+            ->getInterviewAttachments($interviewAttachmentParamHolder);
+        return new EndpointCollectionResult(
+            InterviewAttachmentModel::class,
+            $attachments,
+            new ParameterBag([CommonParams::PARAMETER_TOTAL => count($attachments)])
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:76]:
```suggestion
            new ParameterBag([CommonParams::PARAMETER_TOTAL => $count])
```
this seems wrong. need to get the total from db query

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+    public function getValidationRuleForUpdate(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(
+                self::PARAMETER_ATTACHMENT_ID,
+                new Rule(Rules::POSITIVE)
+            ),
+            new ParamRule(
+                self::PARAMETER_INTERVIEW_ID,
+                new Rule(Rules::ENTITY_ID_EXISTS, [Interview::class])
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:266]: *

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+                    self::PARAMETER_COMMENT,
+                    new Rule(Rules::STRING_TYPE),
+                    new Rule(Rules::LENGTH, [null, self::PARAM_RULE_COMMENT_MAX_LENGTH]),
+                ),
+                true
+            ),
+            new ParamRule(
+                self::PARAMETER_CURRENT_ATTACHMENT,
+                new Rule(Rules::STRING_TYPE),
+                new Rule(Rules::LENGTH, [!null, self::PARAM_RULE_CURRENT_ATTACHMENT_MAX_LENGTH]),
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:279]:
```suggestion
                new Rule(Rules::LENGTH, [null, self::PARAM_RULE_CURRENT_ATTACHMENT_MAX_LENGTH]),
```

File: src/plugins/orangehrmRecruitmentPlugin/Dao/CandidateDao.php [hunk 12]
```diff
@@ -211,4 +212,24 @@ public function saveCandidateHistory(CandidateHistory $candidateHistory): Candid
...
     }
+
+    /**
+     * @param int $candidateId
+     * @return int
+     */
+    public function getInterviewCountByCandidateId(int $candidateId): int
+    {
+        $candidateInterviews = $this->getRepository(Interview::class)->findBy(['candidate'=> $candidateId]);
+        return count($candidateInterviews);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/Dao/CandidateDao.php:223]: better to use query builder. don't load entities to get count

File: src/plugins/orangehrmRecruitmentPlugin/config/routes.yaml [hunk 13]
```diff
@@ -105,6 +105,36 @@ apiv2_recrutiment_candidates_shortlist:
...
+
+apiv2_recruitment_interview_attachments:
+  path: /api/v2/recruitment/interviews/{interviewId}/attachments
+  controller: OrangeHRM\Core\Controller\Rest\V2\GenericRestController::handle
+  methods: [ GET ]
+  defaults:
+    _api: OrangeHRM\Recruitment\Api\InterviewAttachmentAPI
+
+apiv2_recruitment_interviews_attachments:
+  path: /api/v2/recruitment/interviews/attachments
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/config/routes.yaml:132]:
```suggestion
  path: /api/v2/recruitment/interviews/{interviewId}/attachments
```
can't we use this way. you can merge `apiv2_recruitment_interviews_attachments` and  `apiv2_recruitment_interview_attachments`.
there is no restriction to pass all data in request body. can  remove `interviewId` from request body and get from attribute

File: src/plugins/orangehrmRecruitmentPlugin/entity/Decorator/InterviewDecorator.php [hunk 14]
```diff
@@ -82,4 +82,22 @@ private function addInterviewer(Employee $employee): void
...
         $interviewers[] = $employee;
     }
+
+    /**
+     * @return string
+     */
+    public function getInterviewDate(): string
+    {
+        $interviewDate = $this->interview->getInterviewDate();
+        return $interviewDate->format('Y-m-d');
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/entity/Decorator/InterviewDecorator.php:92]:
```suggestion
        return $this->getDateTimeHelper()->formatDateTimeToYmd($interviewDate);
```
can use date tile helper also

File: src/plugins/orangehrmRecruitmentPlugin/entity/Decorator/InterviewDecorator.php [hunk 15]
```diff
@@ -82,4 +82,22 @@ private function addInterviewer(Employee $employee): void
...
+    }
+
+    /**
+     * @return string|null
+     */
+    public function getInterviewTime(): ?string
+    {
+        $interviewTime = $this->interview->getInterviewTime();
+        return !is_null($interviewTime) ? $interviewTime->format('H:i') : null;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmRecruitmentPlugin/entity/Decorator/InterviewDecorator.php:101]:
```suggestion
        return $this->getDateTimeHelper()->formatDateTimeToTimeString($interviewTime);
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Review by RajithaKumara — APPROVED** [review]

File: src/plugins/orangehrmRecruitmentPlugin/Dto/InterviewAttachmentSearchFilterParams.php (hunk 8 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Dto/InterviewAttachmentSearchFilterParams.php:35]: oops, yes

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php (hunk 9 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:76]: fixed

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php (hunk 10 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:266]: done

File: src/plugins/orangehrmRecruitmentPlugin/Dao/CandidateDao.php (hunk 12 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Dao/CandidateDao.php:223]: fixed

File: src/plugins/orangehrmRecruitmentPlugin/config/routes.yaml (hunk 13 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/config/routes.yaml:132]: yes, refactored.

File: src/plugins/orangehrmRecruitmentPlugin/entity/Decorator/InterviewDecorator.php (hunk 14 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/entity/Decorator/InterviewDecorator.php:92]: done

File: src/plugins/orangehrmRecruitmentPlugin/entity/Decorator/InterviewDecorator.php (hunk 15 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/entity/Decorator/InterviewDecorator.php:101]: done

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php (hunk 11 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:279]: done

File: src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php (hunk 6 above)
Comment (SandeepaOHRM) [reply src/plugins/orangehrmRecruitmentPlugin/Api/InterviewAttachmentAPI.php:197]: role based permisson added

**Pull request merged**